From 9c9fd9ae24b45f915f0759d32f662835c26d37c3 Mon Sep 17 00:00:00 2001
From: Dario Binacchi <dario.binacchi@amarulasolutions.com>
Date: Fri, 8 Apr 2022 12:36:33 +0200
Subject: [PATCH] coregrind: fix compilation for uclibc

It fixes a known iusse whose details are described at [1] and more
generally it guarantees that Valgrind is properly compiled for ulibc.

[1] https://www.mail-archive.com/valgrind-users@lists.sourceforge.net/msg05295.html

Suggested-by Michael Trimarchi <michael@amarulasolutions.com>
Co-developed-by: Michael Trimarchi <michael@amarulasolutions.com>
Signed-off-by: Dario Binacchi <dario.binacchi@amarulasolutions.com>
---
 configure.ac                        | 2 ++
 coregrind/m_debuginfo/minilzo-inl.c | 4 ++++
 coregrind/vg_preloaded.c            | 4 ++++
 3 files changed, 10 insertions(+)

diff --git a/configure.ac b/configure.ac
index 6f51dfd9f8fb..c627a9c17a4a 100755
--- a/configure.ac
+++ b/configure.ac
@@ -1204,6 +1204,8 @@ AC_SUBST(DEFAULT_SUPP)
 AC_CHECK_HEADER([features.h])
 
 if test x$ac_cv_header_features_h = xyes; then
+  AC_DEFINE([HAVE_HEADER_FEATURES_H], 1,
+          [Define to 1 if you have the `features.h' header.])
   rm -f conftest.$ac_ext
   cat <<_ACEOF >conftest.$ac_ext
 #include <features.h>
diff --git a/coregrind/m_debuginfo/minilzo-inl.c b/coregrind/m_debuginfo/minilzo-inl.c
index 153d008fd44f..b45ec92eaa85 100644
--- a/coregrind/m_debuginfo/minilzo-inl.c
+++ b/coregrind/m_debuginfo/minilzo-inl.c
@@ -1875,6 +1875,10 @@ extern "C" {
 
 #endif
 
+#ifdef HAVE_HEADER_FEATURES_H
+#include <features.h>
+#endif
+
 #undef LZO_HAVE_CONFIG_H
 #include "minilzo.h"
 
diff --git a/coregrind/vg_preloaded.c b/coregrind/vg_preloaded.c
index 5e2098390847..3809811aede7 100644
--- a/coregrind/vg_preloaded.c
+++ b/coregrind/vg_preloaded.c
@@ -45,6 +45,10 @@
 #include "pub_core_debuginfo.h"  // Needed for pub_core_redir.h
 #include "pub_core_redir.h"      // For VG_NOTIFY_ON_LOAD
 
+#ifdef HAVE_HEADER_FEATURES_H
+#include <features.h>
+#endif
+
 #if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
 
 /* ---------------------------------------------------------------------
-- 
2.32.0

